corehq/tests/test_classcleanup.py

```diff
@@ -0,0 +1,142 @@
+                pass
+
+            @classmethod
+            @maybe_error(self)
+            def tearDownClass(cls):
+                pass
+
+        return [Test()]
+
+    def run_with_errors(self, *errors, error_class=Exception):
+        self.errors = errors
+        self.result = []
+        self.error_class = error_class
+        super().setUp()
+
+    def test_cleanup_in_happy_path(self):
+        self.run_with_errors()
+        eq(self.result, [
+            "setUpClass",
+            "setUp",
+            "runTest",
+            "tearDown",
+            "tearDownClass",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_error_in_set_up_class(self):
+        self.run_with_errors("setUpClass")
+        eq(self.result, [
+            "setUpClass Exception",
+            "cleanup"
+        ])
+
+    def test_cleanup_on_error_in_set_up(self):
+        self.run_with_errors("setUp")
+        eq(self.result, [
+            "setUpClass",
+            "setUp Exception",
+            "tearDownClass",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_error_in_test(self):
+        self.run_with_errors("runTest")
+        eq(self.result, [
+            "setUpClass",
+            "setUp",
+            "runTest Exception",
+            "tearDown",
+            "tearDownClass",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_test_fail(self):
+        self.run_with_errors("runTest", error_class=AssertionError)
+        eq(self.result, [
+            "setUpClass",
+            "setUp",
+            "runTest AssertionError",
+            "tearDown",
+            "tearDownClass",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_error_in_tearDown(self):
+        self.run_with_errors("tearDown")
+        eq(self.result, [
+            "setUpClass",
+            "setUp",
+            "runTest",
+            "tearDown Exception",
+            "tearDownClass",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_error_in_tearDownClass(self):
+        self.run_with_errors("tearDownClass")
+        eq(self.result, [
+            "setUpClass",
+            "setUp",
+            "runTest",
+            "tearDown",
+            "tearDownClass Exception",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_error_in_tearDown_and_tearDownClass(self):
+        self.run_with_errors("tearDown", "tearDownClass")
+        eq(self.result, [
+            "setUpClass",
+            "setUp",
+            "runTest",
+            "tearDown Exception",
+            "tearDownClass Exception",
+            "cleanup",
+        ])
+
+
+def maybe_error(test):
+    def decorator(func):
+        @wraps(func)
+        def wrapper(self):
+            func(self)
+            test.result.append(func.__name__)
+            if func.__name__ in test.errors:
+                test.result[-1] += f" {test.error_class.__name__}"
+                raise test.error_class
+        return wrapper
+    return decorator
```

Comment:
I found this decorator a bit confusing, as it's doing two things at once -- it's responsible for logging execution, as well as for raising an error, but it's name reflects only the latter responsibility (and it seems like its primary responsibility is the former). Not a blocker, but I could see this being two decorators -- a logger, which logs the path and whether or not it received an exception, and a raiser, which triggers the exception itself. Just reading over `makeSuite`, it wasn't intuitive to me why the methods were decorated with `maybeError`

Reply:
https://github.com/dimagi/commcare-hq/pull/31084/commits/fbae216d5c9be5ff0a92871bd1f9493d6f6ee8cd

I renamed it to hopefully make it clearer, but didn't split it into two separate decorators because that would either require more logic to protect the logger from errors or tightly couple the decorators by adding significance the order in which they are applied. Also two decorators felt like superfluous boilerplate.